feat: Editorial review: Attribution reporting API (#33172)
mdn/content: files/en-us/web/api/attribution_reporting_api/registering_sources/index.md

---
title: Registering attribution sources
slug: Web/API/Attribution_Reporting_API/Registering_sources
page-type: guide
status:
  - experimental
---

{{SeeCompatTable}}{{DefaultAPISidebar("Attribution Reporting API")}}

> This article explains how to register attribution sources.

Reviewer: I know that we have context in breadcrumb, but I'm a fan of spelling it out. So something like

```suggestion
This article explains how to register attribution sources when using the [Attribution Reporting API](/en-US/docs/Web/API/Attribution_Reporting_API).
An attribution source is an [ad ?] that you want to measure interactions with, that may be represented by a link, image or script ....
```

## Basic methodology

Attribution sources take the form of links, images, or scripts contained within content that you want to measure interactions with (for example, they might be ads that you want to measure conversions on). These cause the browser to store source data in a private local cache (accessible only by the browser) when specific actions occur. The different attribution source types are registered and signal actions in different ways — they are differentiated as:

- Navigation sources, which signal in response to navigation — for example when the user clicks on a link or activates it with the keyboard, or when a navigation occurs as a result of a {{domxref("Window.open()")}} call. See [Navigation-based attribution sources](#navigation-based_attribution_sources) for examples.

> - Event sources, which signal in response to events firing. See [Event-based attribution sources](#event-based_attribution_sources) for examples.

Reviewer: When you say "signal", do you mean "trigger an attribution" or "register a source" or something else? 

It would be nice to understand the difference - I mean I understand what a navigation is, so I am envisaging that the signal is perhaps a source registration on clicking a link to go to the ad website?

What's an event though - how does that map to some kind of user behaviour?

Author: I've replaced signal with "cause the browser to store source data" in both cases — this is what we are signalling.

An event is just a regular DOM event - a `load` event in the case of the img or script, or any custom event you want to invoke the fetch request, in the case of JavaScript-based event sources.

However, what happens behind the scenes to register sources and retrieve and store the source data is the same in all cases.

> 1. All of the sources send an {{httpheader("Attribution-Reporting-Eligible")}} header on a request, which indicates that the response is eligible to register a source. For example:

Reviewer: "header on a request" - on a request to who?

Author: Updated to "...header on a request to the server measuring the interactions (typically the advertiser's server)"

   - The following example will cause an [event-level report](/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#event-level_reports) to be generated when a [trigger](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_triggers) is matched to a source:

     ```js
     res.set(
       "Attribution-Reporting-Register-Source",
       JSON.stringify({
         source_event_id: "412444888111012",
         destination: "https://advertiser.example",
         expiry: "604800",
         priority: "100",
         debug_key: "122939999",
         event_report_window: "86400",
       }),
     );
     ```

     The only required field in this context is `destination`, which specifies 1–3 sites on which a trigger is expected to occur. These are used to match the attribution trigger to the source when a trigger is interacted with. The other fields specified above do the following:

>      - `"source_event_id"`: A string representing an ID for the attribution source, which can be used to map it to other information when the attribution source is interacted with, or aggregate information at the reporting endpoint (see [Generating reports > Basic process](/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#basic_process) for endpoint information).

Reviewer: How is this generated? I.e. is this ad-specific, or UI element specific. Does it matter?

Author: It can be any string — it is up to the developer. I'm not sure we need to specify this?

     - `"event_report_window"`: A string representing a time in seconds, after which subsequent triggers won't be attributable to this source for the purpose of producing event-level reports.

     See {{httpheader("Attribution-Reporting-Register-Source")}} for a detailed description of all the available fields.

   - To make the browser generate a [summary report](/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#summary_reports) when a trigger is matched to a source, you need to include some extra fields, _in addition_ to those required for generating an event-level report.

     ```js
     res.set(
       "Attribution-Reporting-Register-Source",
       JSON.stringify({
         source_event_id: "412444888111012",
         destination: "https://advertiser.example",
         expiry: "604800",
         priority: "100",
         debug_key: "122939999",
         event_report_window: "86400",

         aggregation_keys: {
           campaignCounts: "0x159",
           geoValue: "0x5",
         },
         aggregatable_report_window: "86400",
       }),
     );
     ```

     The additional fields in this example are:

>      - `"aggregation_keys"`: An object containing user-provided keys representing different data points to aggregate report values under.

Reviewer: At this point having read this and the register source header docs, I don't know how I would choose the key values. I assume this is covered in the reporting doc.

Reviewer: Yes. I guess optional here.

     - `"aggregatable_report_window"`: A string representing a time in seconds after which trigger data will no longer be included in generated aggregatable reports.

     Again, see {{httpheader("Attribution-Reporting-Register-Source")}} for a detailed description of all the available fields.

3. When the specified action occurs on the registered attribution source, the browser stores the provided source data in its private local cache.

## Navigation-based attribution sources

> There are a couple of different types of navigation-based attribution sources that can be registered — those based on HTML (which use the `attributionsrc` attribute) and those based on {{domxref("Window.open()")}} calls (which use an `attributionsrc` window feature).

Reviewer: It might be worth repeating what Navigation sources "correspond to". Clicks on ads I think.

Author: Updated to "...navigation-based attribution sources (for example, clicking on an ad) that can be registered..."

Does that work?

### HTML-based navigation sources

> To register a navigation-based attribution source you can add the `attributionsrc` attribute to an appropriate {{htmlelement("a")}} element, either declaratively:

Reviewer: 1. It is worth saying here what this does/sets - that is, the server where the registration headers are sent, and from which the responses for registration are expected. Where is this server normally? I.e. is it the server that hosts the ads?
2. Perhaps one example showing setting an explicit source and explaining that you can specify some other source. Right now my assumption would be that you probably have to use href, since that's the only URL shown

Reviewer: Ah, I see you cover this at the section specifying_urls_inside_attributionsrc - I still think a word or two here about what is being specified makes sense, followed by a link  down

Author: OK, fair call. I've updated this section with an extra little paragraph:

> If you leave the attribute value blank, the registration request will be sent to the location being linked to. It is also possible to specify one or more additional URLs inside the value to send the registration request to; see the [Specifying URLs inside attributionsrc](#specifying_urls_inside_attributionsrc) for more details.

And I've added an equivalent paragraph to event-based section, and to the triggers guide in the appropriate place.

```html
<a href="https://shop.example" attributionsrc target="_blank">
  Click to visit our shop
</a>
```

> Or via the {{domxref("HTMLAnchorElement.attributionSrc")}} property:

Reviewer: 1. Can I unregister? (or would I never want to?)
2. Perhaps one example showing setting a target?

Author: AFAIK, there is no way to unregister a source or trigger dynamically after registration. I don't think there's a need to do this; you'd set up a campaign, leave it to run, and then update your code to change or remove it when required.

Reviewer 2: The closest thing to this is [ARA's integration with `Clear-Site-Data`](https://wicg.github.io/attribution-reporting-api/#clear-site-data-integration).

But there's no way to delete individual sources or triggers or reports.

Author: @apasel422 does the Clear-Site-Data integration have its own separate chromestatus.com entry, or was it just added along with the general API implementation? I'm thinking that it'd be good to mention this integration somewhere in the ARA docs, and mention it on the Clear-Site-Data page. But I'd like to do that as a follow-up bit of work, rather than add to the complexity of this PR.

Reviewer 2: No, it doesn't have its own entry. I agree that we should mention it, but it's fine to do that in a separate PR.

### Window.open()-based navigation sources

You can also add the `attributionsrc` feature keyword to the features property of a {{domxref("Window.open()")}} call. In this example we run it in response to a `click` event being fired:

```js
elem.addEventListener("click", () => {
  window.open("https://shop.example", "_blank", "attributionsrc");
});
```

> **Note:** When setting up a [`click`](/en-US/docs/Web/API/Element/click_event) event like in the above example, it is advisable to set it on a control where a click is expected, such as a {{htmlelement("button")}} or {{htmlelement("a")}} element. This makes more sense semantically, and is more accessible to both screenreader and keyboard users.

> **Note:** To register an attribution source via `open()`, it must be called with [transient activation](/en-US/docs/Glossary/Transient_activation) (i.e. inside a user interaction event handler such as `click`) within five seconds of user interaction.

## Event-based attribution sources

> There are a couple of different types of event-based attribution sources that can be registered — those based on HTML (which use the `attributionsrc` attribute like we saw above with `<a>` element) and those based on JavaScript.

Reviewer: As above, would be nice to state what kind of interaction these "correspond to", contrasted to navigation, which is a link to the ad site (I think)

Author: I've rewritten this opening paragraph to be a bit more descriptive:

> Event-based attribution sources cause the browser to store source data in response to some kind of event firing, such as the `load` event in the case of an `<img>` or `<script>` element (which use the `attributionsrc` attribute like we saw above with the `<a>` element), or a custom event of your choice set in your JavaScript.

### HTML-based event sources

To register an event-based attribution source via HTML, you can add the `attributionsrc` attribute to an appropriate element — {{htmlelement("img")}} or {{htmlelement("script")}}.

Let's look at an `<img>` element example:

```html
<img src="advertising-image.png" attributionsrc />
```

> You could also achieve this via the {{domxref("HTMLImageElement.attributionSrc")}} property:

Reviewer: Again, perhaps a line stating "as in the navigation sources above, the `attributionSrc` can also indicate where the registration headers are to be sent"

```js
const imgElem = document.querySelector("img");
imgElem.attributionSrc = "";
```

> The browser stores the source data associated with the source (as provided in the {{httpheader("Attribution-Reporting-Register-Source")}} response header) when the browser receives the response containing the image file. Bear in mind that users might not necessarily be able to perceive the image at all — it might be a 1x1 transparent tracking pixel that is only being used for attribution reporting.

Reviewer: This is why we need to explain what the event triggers correspond to. Because then it makes sense of the fact you have an image in your page that is somehow doing something but which the user can't interact with. I think that might help me understand WHEN you would use an image too. It would be good to spell that out.

Right now my assumption is that an image is used when, say, you want to register ads without user interaction. So you might have a bunch of these little images embedded in the page, one per ad, but invisible to the user. When the user loads the page the images get downloaded for all the adds in the page, and the registration headers get sent for each of them.

This contrasts with using an explicit user trigger, such as a navigation.

Author: I've updated all the sections to say when the interactions occur, causing the browser to store the source data, in each case. Have another look and let me know if you think it needs any more detail.

A {{htmlelement("script")}} example might look like so:

```html
<script src="advertising-script.js" attributionsrc />
```

Or via the {{domxref("HTMLScriptElement.attributionSrc")}} property:

```js
const scriptElem = document.querySelector("script");
scriptElem.attributionSrc = "";
```

In this case, the browser stores the source data associated with the source (as provided in the {{httpheader("Attribution-Reporting-Register-Source")}} response header) when the browser receives the response containing the script.

### JavaScript-based event sources

> Script-based attribution sources are more versatile than HTML-based attribution sources. You can set up a script to initiate a request that is eligible to register an attribution source based on whatever request suits your app.

Reviewer: As for navigation/images, would be good to have a concrete example of when you might use this approach

Author: Down the bottom of the section, I've added a line:

> This is a flexible approach useful when you want to store source data in response to custom interactions, for example, clicking a custom element or submitting a form.

Author: Is this the kind of thing you are after for the other source types?

To set up a script-based attribution source, you can:

> 1. Send a {{domxref("fetch()")}} request containing the `attributionReporting` option:

Reviewer: These are either/or, not a list. I'd bullet at least, and perhaps prefix with "either"

Author: Good call. I've done both, here and on the triggers guide page.

       req.setAttributionReporting(attributionReporting);
       req.send();
     } else {
       throw new Error("Attribution reporting not available");
       // Include recovery code here as appropriate
     }
   }

   // Associate the interaction trigger with whatever
   // event makes sense for your code (does not have to be a
   // DOM event/user interaction)
   elem.addEventListener("click", triggerSourceInteraction);
   ```

> **Note:** The request can be for any resource. It doesn't need to have anything directly to do with the Attribution Reporting API, and can be a request for JSON, plain text, an image blob, or whatever else makes sense for your app.

## Specifying URLs inside attributionsrc

So far, in all the examples we have seen, the `attributionsrc` attribute/feature or `attributionSrc` property has been left blank, taking the value of an empty string. This is fine if the server that holds the requested resource is the same server that you also want to handle the registration, i.e. receive the {{httpheader("Attribution-Reporting-Eligible")}} header and respond with the {{httpheader("Attribution-Reporting-Register-Source")}} header.

> However, it might be the case that the requested resource is not on a server you control, and you want to register the attribution source via a separate server that you _do_ control. In this case, you can specify one or more URLs as the value of `attributionsrc`. When the resource request occurs, the {{httpheader("Attribution-Reporting-Eligible")}} header will be sent to the URL(s) specified in `attributionsrc` in addition to the resource origin; these URLs can then respond with the {{httpheader("Attribution-Reporting-Register-Source")}} to register the source.

Reviewer: P.S. If you are NOT going to explain "why" in more than this detail, I'd embed all this above rather than pulling out into its own section.

> However, it might be the case that the requested resource is not on a server you control, and you want to register the attribution source via a separate server that you _do_ control. In this case, you can specify one or more URLs as the value of `attributionsrc`. When the resource request occurs, the {{httpheader("Attribution-Reporting-Eligible")}} header will be sent to the URL(s) specified in `attributionsrc` in addition to the resource origin; these URLs can then respond with the {{httpheader("Attribution-Reporting-Register-Source")}} to register the source.

Author: I've updated the text to:

> However, it might be that the requested resource is not on a server you control, or you just want to handle registering the attribution trigger on a different server. In such cases, ...

This downplays the "control" aspect that you seem to be getting stuck on.

In terms of the control example, I'd imagine that you might end up in a situation where you want to register as a source some kind of ad banner image that is being served from a third-party ad provider. They won't send back the necessary response headers to complete registration, so instead, you put your own server's URL inside `attributionsrc`.

Do you think I need to explain that example on the page?

  Click to visit our shop
</a>
```

Or in JavaScript via the `attributionSrc` property:

```js
// encode the URLs in case they contain special characters
// such as '=' that would be improperly parsed.
const encodedUrlA = encodeURIComponent("https://a.example/register-source");
const encodedUrlB = encodeURIComponent("https://b.example/register-source");

const aElem = document.querySelector("a");
aElem.attributionSrc = `${encodedUrlA} ${encodedUrlB}`;
```

In the case of a {{domxref("Window.open()")}} call, the different URLs would have to be listed as multiple separate `attributionsrc` features in the [`windowFeatures`](/en-US/docs/Web/API/Window/open#windowfeatures) parameter, separated by commas or whitespace:

```js
// encode the URLs in case they contain special characters
// such as '=' that would be improperly parsed.
const encodedUrlA = encodeURIComponent("https://a.example/register-source");
const encodedUrlB = encodeURIComponent("https://b.example/register-source");

elem.addEventListener("click", () => {
  window.open(
    "https://ourshop.example",
    "_blank",
    `attributionsrc=${encodedUrlA},attributionsrc=${encodedUrlB}`,
  );
});
```

> > **Note:** Specifying multiple URLs at once means that multiple attribution source events can be registered for the same attribution source. You might for example have different campaigns that you are trying to measure the success of, which involve generating different reports on different data.

Reviewer: This is the first mention of the second point. Should probably also be where this is defined, such as https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/HTMLImageElement/attributionSrc

Author: Good call. I've added this in all the relevant places. I also changed the wording a bit, to suit my rewrite on the guide pages:

> **Note:** Specifying multiple URLs means that multiple attribution sources can be registered on the same feature. You might for example have different campaigns that you are trying to measure the success of, which involve generating different reports on different data.

This article explains how to register attribution sources when using the [Attribution Reporting API](/en-US/docs/Web/API/Attribution_Reporting_API).

## Basic methodology

Attribution sources take the form of links, images, or scripts contained within content that you want to measure interactions with (for example, they might be ads that you want to measure conversions on). These cause the browser to store source data in a private local cache (accessible only by the browser) when specific actions occur. The different attribution source types are registered and signal interactions in different ways — they are differentiated as:

- Navigation sources, which cause the browser to store source data in response to navigation — for example when the user clicks on a link or activates it with the keyboard, or when a navigation occurs as a result of a {{domxref("Window.open()")}} call. See [Navigation-based attribution sources](#navigation-based_attribution_sources) for examples.
- Event sources, which cause the browser to store source data in response to events firing. See [Event-based attribution sources](#event-based_attribution_sources) for examples.

> However, what happens behind the scenes to register sources and retrieve and store the source data is the same in all cases.

Reviewer: ```suggestion
However, what happens behind the scenes to register sources and retrieve and store the source data, is the same in both cases.
```

1. All of the sources send an {{httpheader("Attribution-Reporting-Eligible")}} header on a request to the server measuring the interactions (typically the advertiser's server), which indicates that the response is eligible to register a source. For example:

   ```http
   Attribution-Reporting-Eligible: navigation-source
   ```

> 2. When the server receives a request that includes an `Attribution-Reporting-Eligible` header, it can include an {{httpheader("Attribution-Reporting-Register-Source")}} header along with the response. Its value is a JSON string that provides the information the browser should store when the attribution source is interacted with. The information you include in this header also determines which types of report the browser will generate:

Reviewer: This is out of date terminology for when we thought that you'd register a source, then trigger it. How about:
```suggestion
2. When the server receives a request that includes an `Attribution-Reporting-Eligible` header, it can include an {{httpheader("Attribution-Reporting-Register-Source")}} header along with the response. Its value is a JSON string that provides the information the browser should store about the attribution source that was interacted with. The information included in this header also determines which types of report the browser will generate:
```

> 2. When the server receives a request that includes an `Attribution-Reporting-Eligible` header, it can include an {{httpheader("Attribution-Reporting-Register-Source")}} header along with the response. This takes a JSON string as its value that provides the information the browser should store when the attribution source is interacted with. The information you include in this header also determines which types of report the browser will generate:

Reviewer: Here and in trigger doc, is "takes a JSON" good?

```suggestion
2. When the server receives a request that includes an `Attribution-Reporting-Eligible` header, it can include an {{httpheader("Attribution-Reporting-Register-Source")}} header along with the response. This has a JSON string as its value that provides the information the browser should store when the attribution source is interacted with. The information you include in this header also determines which types of report the browser will generate:
```

Author: Meh, this was horrible wording. I've updated it to "Its value is a JSON string that provides...", and the equivalent in the triggers guide.

   - The following example will cause an [event-level report](/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#event-level_reports) to be generated when a [trigger](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_triggers) is matched to a source:

>      ```js

Reviewer: So this does not mention ["trigger_data_matching"](https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) or ["trigger_data"](https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data).

You kind of have to mention the data matching here, because otherwise you'll get modulus, which you probably don't want.

Author: OK, `"trigger_data_matching"` and `"trigger_data"` included in the examples, both here and in the HTTP header ref page, and explanations added to this page.

     res.set(
       "Attribution-Reporting-Register-Source",
       JSON.stringify({
         source_event_id: "412444888111012",
         destination: "https://advertiser.example",
         expiry: "604800",
         priority: "100",
         debug_key: "122939999",
         event_report_window: "86400",
       }),
     );
     ```

     The only required field in this context is `destination`, which specifies 1–3 sites on which a trigger is expected to occur. These are used to match the attribution trigger to the source when a trigger is interacted with. The other fields specified above do the following:

>      - `"source_event_id"`: A string representing an ID for the attribution source, which can be used to map it to other information when the attribution source is interacted with, or aggregate information at the reporting endpoint (see [Generating reports > Basic process](/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#basic_process) for endpoint information).

Reviewer: If you think you'll ever need to link to these, consider making them definition lists

Author: When linking to definitions of these fields, I always link to the ones on the header page anyway, where they are presented as `<dl>`s.

     ```js
     res.set(
       "Attribution-Reporting-Register-Source",
       JSON.stringify({
         source_event_id: "412444888111012",
         destination: "https://advertiser.example",
         expiry: "604800",
         priority: "100",
         debug_key: "122939999",
         event_report_window: "86400",

         aggregation_keys: {
           campaignCounts: "0x159",
           geoValue: "0x5",
         },
         aggregatable_report_window: "86400",
       }),
     );
     ```

     The additional fields in this example are:

     - `"aggregation_keys"`: An object containing user-provided keys representing different data points to aggregate report values under.
     - `"aggregatable_report_window"`: A string representing a time in seconds after which trigger data will no longer be included in generated aggregatable reports.

     Again, see {{httpheader("Attribution-Reporting-Register-Source")}} for a detailed description of all the fields available on this header.

> 3. When the user interacts with the attribution source, the browser stores the provided source data in its private local cache.

Reviewer: See my point above - the registration occurs because there has been an interaction with the source.

```suggestion
```

Author: Ah yes, these descriptions are a bit wrong in terms of what we've said on the main page. I've updated points 1-3, so that they are more correct.

});
```

In this case, the interaction occurs and the browser stores the source data when `Window.open()` is invoked, and the browser receives the response.

> **Note:** When setting up a [`click`](/en-US/docs/Web/API/Element/click_event) event like in the above example, it is advisable to set it on a control where a click is expected, such as a {{htmlelement("button")}} or {{htmlelement("a")}} element. This makes more sense semantically, and is more accessible to both screenreader and keyboard users.

> **Note:** To register an attribution source via `open()`, it must be called with [transient activation](/en-US/docs/Glossary/Transient_activation) (i.e. inside a user interaction event handler such as `click`) within five seconds of user interaction.

## Event-based attribution sources

Event-based attribution sources cause the browser to store source data in response to some kind of event firing, such as the `load` event in the case of an `<img>` or `<script>` element (which use the `attributionsrc` attribute like we saw above with the `<a>` element), or a custom event of your choice set in your JavaScript.

### HTML-based event sources

> HTML-based event sources can be used for measuring interactions with a publisher's page when it first loads — or more precisely when a tracking pixel `<img>` or specific `<script>` loads. To register an event-based attribution source via HTML, you can add the `attributionsrc` attribute to an appropriate element — {{htmlelement("img")}} or {{htmlelement("script")}}.

Reviewer: FWIW not sure whether we need to be precise about the tracking pixel. That's a main use case, but it wouldn't have to be. Probably OK. 
 
```suggestion
HTML-based event sources can be used for measuring interactions with a publisher's page when it loads — or more precisely when a tracking pixel `<img>` or specific `<script>` loads. To register an event-based attribution source via HTML, you can add the `attributionsrc` attribute to an appropriate element — {{htmlelement("img")}} or {{htmlelement("script")}}.
```

If you leave the attribute value blank, the registration request will be sent to the server the requested resource is hosted on. It is also possible to specify one or more additional URLs inside the value to send the registration request to; see [Specifying URLs inside attributionsrc](#specifying_urls_inside_attributionsrc) for more details.

Let's look at an `<img>` element example:

```html
<img src="advertising-image.png" attributionsrc />
```

You could also achieve this via the {{domxref("HTMLImageElement.attributionSrc")}} property:

```js
const imgElem = document.querySelector("img");
imgElem.attributionSrc = "";
```

> In this case, the interaction occurs and the browser stores the source data when the browser receives the response containing the image file (i.e. when the `load` event occurs). Bear in mind that users might not necessarily be able to perceive the image at all — it might be a 1x1 transparent tracking pixel that is only being used for attribution reporting.

Reviewer: ```suggestion
The browser stores the attribution source data when the browser receives the response containing the image file (i.e. when the `load` event occurs). Bear in mind that users might not necessarily be able to perceive the image at all — it might be a 1x1 transparent tracking pixel that is only being used for attribution reporting.
```

Author: Good call; changed.

     Again, see {{httpheader("Attribution-Reporting-Register-Source")}} for a detailed description of all the fields available on this header.

3. When the user interacts with the attribution source, the browser stores the provided source data in its private local cache.

## Navigation-based attribution sources

Navigation sources are useful for measuring interactions with links — for example, a user may see an ad on a publisher's page, and click it to navigate to the advertiser's page where a conversion will hopefully occur.

There are a couple of different types of navigation-based attribution sources (for example, clicking on an ad) that can be registered — those based on HTML (which use the `attributionsrc` attribute) and those based on {{domxref("Window.open()")}} calls (which use an `attributionsrc` window feature).

### HTML-based navigation sources

To register a navigation-based attribution source you can add the `attributionsrc` attribute to an appropriate {{htmlelement("a")}} element, which specifies where the registration request will be sent.

> If you leave the attribute value blank, the registration request will be sent to the location being linked to. It is also possible to specify one or more additional URLs inside the value to send the registration request to; see the [Specifying URLs inside attributionsrc](#specifying_urls_inside_attributionsrc) for more details.

Reviewer: FWIW yes, this fixes my previous comments

Author: Good good ;-)

      // Include recovery code here as appropriate
    }
  }

  // Associate the interaction trigger with whatever
  // event makes sense for your code (does not have to be a
  // DOM event/user interaction)
  elem.addEventListener("click", triggerSourceInteraction);
  ```

In this case, the interaction occurs and the browser stores the source data when the browser receives the response from the fetch request.

> **Note:** The request can be for any resource. It doesn't need to have anything directly to do with the Attribution Reporting API, and can be a request for JSON, plain text, an image blob, or whatever else makes sense for your app.

## Specifying URLs inside attributionsrc

So far, in all the examples we have seen, the `attributionsrc` attribute/feature or `attributionSrc` property has been left blank, taking the value of an empty string. This is fine if the server that holds the requested resource is the same server that you also want to handle the registration, i.e. receive the {{httpheader("Attribution-Reporting-Eligible")}} header and respond with the {{httpheader("Attribution-Reporting-Register-Source")}} header.

> However, it might be that the requested resource is not on a server you control, or you just want to handle registering the attribution source on a different server. In such cases, you can specify one or more URLs as the value of `attributionsrc`. When the resource request occurs, the {{httpheader("Attribution-Reporting-Eligible")}} header will be sent to the URL(s) specified in `attributionsrc` in addition to the resource origin; these URLs can then respond with the {{httpheader("Attribution-Reporting-Register-Source")}} to register the source.

Reviewer: So  if you use this in an img you might actually get back multiple images for one image element?

Author: No; only the original image request will return an image. The responses to the requests to the `attributionsrc` URLs will not. They should only return the `Attribution-Reporting-Register-Source`.
